keps/sig-auth/2718-20210511-client-exec-proxy/README.md

```diff
@@ -0,0 +1,854 @@
+posture of the exec credential feature present in k8s.io/client-go today.
+
+The security and UX will be reviewed by SIG Auth.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Proxy Proposal
+
+The proxy exec plugin is a long-running proxy process which is executed by the
+client on first use.  When executed, the exec proxy plugin will bind and listen
+on a socket, and will a URL on stdout, over which the client would proxy all
+API server bound requests.  The proxy binary location is defined in the
+`ExecConfig` section of a kubeconfig `user`.
+
+The proxy plugin process will be executed from client-go the first time it is
+needed, on transport setup.  If the client fails to connect to the socket
+because no one is listening, client-go will reattempt to exec the plugin.
+Client-go will support both plugins binaries which are themselves the proxy
+process, and additionally plugin binaries which complete, but are responsible
+for starting a proxy process and returning a URL to the proxy.
+
+The client will send its request to the proxy plugin process, and the proxy
+will forward the request to the API server, rather than returning the modified
+request to the client to be sent (see alternatives considered).  The downside
+of this approach is that the proxy must duplicate the some of the transport
+functionality of client packages, like respecting existing proxy configuration,
+but allows for more flexibility for proxy plugin developers.
+
+### Authentication to Proxy
+
+When communicating to the proxy over TCP/IP, the authentication method to the
+proxy is certificate based. The proxy passes back a client certificate and
+certificate authority that the client can use on connection.
+
+When communicating to the proxy over a unix domain socket, the authentication
+method to the proxy is file permission based.  The GID owner of the client
+process is passed to the proxy over stdin, which the proxy should give
+read/write permissions on the socket to.
+
+### Caching
+
+The connection from the client to the proxy can be cached in the client based
+on the parameters in the ExecProxyRequest: the proxyURL and cluster details.
+
+### Proxy Shutdown
+
+In the case where the client dies unexpectedly, due to a crash or SIGKILL, the
+proxy should also die.  This could be accomplished with 2 approaches: the proxy
+could quit after a certain amount of idle time, or the proxy could attempt to
+take a file lock that the parent holds and quit if it ever successfully takes
+the lock.  Both are reasonable approaches, so idle quit will be used due to its
+simplicity.  Proxies that are idle for more than 5 minutes will be required to
+quit.
+
+### API
```

Comment:
I disagree with effectively the entire structure of this API.  I expect this to be an extension to the existing client-go credential plugin API, specifically as extension to `ExecCredentialStatus`.

Reply:
Moved to existing `ExecCredential` API.